### ISSUE TITLE
Flex Attention runs slower than FlashAttention-2

### ISSUE THREAD

**Issue by Yangchenchen-Jin** [meta-pytorch/attention-gym#172]
Hi, 

I’ve been benchmarking `flex_attention` against `flash_attention_2` using the official `ALL_ATTENTION_FUNCTIONS` API from `transformers`.

While Flex Attention is expected to be more general and efficient for sparse or structured masks, it appears **slower** than FlashAttention-2.

Below are the details, benchmark results, and environment information.

## Environment
- OS: Red Hat Enterprise Linux 9.5 (x86_64) 
- Python: 3.12.11 (conda-forge)
- PyTorch: 2.8.0+cu128
- CUDA used to build PyTorch: 12.8
- cuDNN: 9.10.2.21
- Triton: 3.4.0
- GPU: NVIDIA B200 (2 GPUs)
- Driver version: 570.148.08
- dtype: bfloat16
- Transformers: 4.46.2
- FlashAttention: 2.5.8  
(collected via `python -m torch.utils.collect_env`)

## Reproducible code
```python
import math, time, torch, itertools
from types import SimpleNamespace
from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS
from transformers.utils import is_flash_attn_2_available

try:
    # Flex block mask API (used only for flex_attention)
    from torch.nn.attention.flex_attention import create_block_mask
    _flex_available = True
except Exception:
    _flex_available = False

# ========================= Basic Setup =========================
torch.set_float32_matmul_precision("high")
torch.backends.cuda.matmul.allow_tf32 = True
device  = "cuda" if torch.cuda.is_available() else "cpu"
dtype   = (torch.bfloat16 if (device=="cuda" and torch.cuda.is_bf16_supported()) else
           (torch.float16 if device=="cuda" else torch.float32))

# Benchmark scales: (name, B, Hq, Hk, T, Dh)
SCALES = [
    ("small",   1,  2,  1,   64,  16),   # Dh>=16 to satisfy flex_attention requirement
    ("medium",  1,  8,  4,  1024,  64),
    ("large",   1, 16, 8,  4096,  64),
]
IS_CAUSAL_OPTIONS = [True, False]
REPEATS_BY_T = lambda T: (300 if T<=64 else (100 if T<=1024 else 30))
WARMUP = 5

# Mask types (only used by flex_attention)
# Each mask kind is passed to create_block_mask for (B, Hq, QLEN, KLEN)
FLEX_MASKS = {
    "none": None,
    "causal": "causal",          # causal mask: keep = (q >= kv)
    "diag_lastrow": "diag_lastrow",  # keep diagonal + last row
    "diag": "diag",              # keep only diagonal
}

# ========================= Helper Functions =========================
def make_qkv(B,Hq,Hk,T,Dh, dtype, device):
    # Deterministic small random tensors for consistent comparison
    g = torch.Generator(device=device).manual_seed(42)
    q = torch.rand((B,Hq,T,Dh), dtype=dtype, device=device, generator=g) * 0.01
    k = torch.rand((B,Hk,T,Dh), dtype=dtype, device=device, generator=g) * 0.01
    v = torch.rand((B,Hk,T,Dh), dtype=dtype, device=device, generator=g) * 0.01
    return q.contiguous(), k.contiguous(), v.contiguous()

def _time_once(fn):
    if device == "cuda":
        start = torch.cuda.Event(enable_timing=True)
        end   = torch.cuda.Event(enable_timing=True)
        torch.cuda.synchronize()
        start.record()
        out = fn()
        end.record()
        torch.cuda.synchronize()
        return out, start.elapsed_time(end)/1000.0
    else:
        t0 = time.perf_counter(); out = fn(); t1 = time.perf_counter()
        return out, (t1 - t0)

def _flex_build_block_mask(mask_kind, B,Hq,QLEN,KLEN, device):
    if mask_kind is None:
        return None

    def _mask_mod_factory(kind):
        def _mask_mod(b, h, q, kv):
            if not isinstance(q, torch.Tensor):  q_ = torch.tensor(q, device=device)
            else:                                q_ = q
            if not isinstance(kv, torch.Tensor): kv_ = torch.tensor(kv, device=device)
            else:                                kv_ = kv
            keep_diag     = (q_ == kv_)
            keep_last_row = (q_ == (QLEN - 1))
            if kind == "causal":
                keep = (q_ >= kv_)
            elif kind == "diag_lastrow":
                keep = (keep_diag | keep_last_row)
            elif kind == "diag":
                keep = keep_diag
            else:
                raise ValueError(f"Unknown mask kind: {kind}")
            return keep
        return _mask_mod

    _mask_mod = _mask_mod_factory(mask_kind)
    return create_block_mask(_mask_mod, B, Hq, QLEN, KLEN, device=device)

def bench_one(fn, fake_self, q,k,v, attention_mask, scaling, repeats, warmup, is_causal):
    # Warmup
    for _ in range(warmup):
        _ = fn(fake_self, q,k,v, attention_mask, dropout=0.0, scaling=scaling, is_causal=is_causal)
    if device == "cuda":
        torch.cuda.synchronize()

    # Timing
    total = 0.0
    for _ in range(repeats):
        def _call():
            return fn(fake_self, q,k,v, attention_mask, dropout=0.0, scaling=scaling, is_causal=is_causal)
        _, dt = _time_once(_call)
        total += dt
    return total / repeats

# ========================= Main Benchmark Loop =========================
rows = []
for (name, B,Hq,Hk,T,Dh), is_causal in itertools.product(SCALES, IS_CAUSAL_OPTIONS):
    if Dh < 16:
        continue

    q,k,v = make_qkv(B,Hq,Hk,T,Dh, dtype, device)
    scaling = 1.0 / math.sqrt(Dh)
    repeats = REPEATS_BY_T(T)
    fake_self = SimpleNamespace(training=False, is_causal=is_causal)

    backends = ["flash_attention_2", "flex_attention"]
    for be in backends:
        fn = ALL_ATTENTION_FUNCTIONS.get(be)
        if fn is None:
            rows.append((name, B,Hq,Hk,T,Dh, is_causal, be, "-", "N/A: not found"))
            continue

        if be == "flash_attention_2":
            if device != "cuda":
                rows.append((name,B,Hq,Hk,T,Dh,is_causal,be,"-", "Skip: need CUDA"))
                continue
            if not is_flash_attn_2_available():
                rows.append((name,B,Hq,Hk,T,Dh,is_causal,be,"-", "Skip: flash-attn not available"))
                continue

        # Multiple masks for flex_attention, only "none" for others
        mask_list = (FLEX_MASKS.items() if (be=="flex_attention" and _flex_available) else [("none", None)])

        for mask_name, mask_kind in mask_list:
            if be != "flex_attention":
                attn_mask = None
            else:
                attn_mask = (_flex_build_block_mask(mask_kind, B,Hq,T,T, device) 
                             if mask_kind is not None else None)

            try:
                avg_s = bench_one(fn, fake_self, q,k,v, attn_mask, scaling, repeats, WARMUP, is_causal)
                rows.append((name,B,Hq,Hk,T,Dh, is_causal, be, mask_name, f"{avg_s*1e3:.3f} ms"))
            except Exception as e:
                rows.append((name,B,Hq,Hk,T,Dh, is_causal, be, mask_name, f"Skip: {type(e).__name__}: {e}"))

# ========================= Print Results =========================
header = ["scale","B","Hq","Hk","T","Dh","is_causal","backend","mask(flex-only)","avg_time"]
colw = [max(len(str(x)) for x in [h] + [r[i] for r in rows]) for i,h in enumerate(header)]

print("\n" + " | ".join(h.ljust(colw[i]) for i,h in enumerate(header)))
print("-" * (sum(colw)+3*(len(header)-1)))
for r in rows:
    print(" | ".join(str(r[i]).ljust(colw[i]) for i in range(len(header))))
```

## Results
```python
scale  | B | Hq | Hk | T    | Dh | is_causal | backend           | mask(flex-only) | avg_time
---------------------------------------------------------------------------------------------
small  | 1 | 2  | 1  | 64   | 16 | True      | flash_attention_2 | none            | 0.061 ms
small  | 1 | 2  | 1  | 64   | 16 | True      | flex_attention    | none            | 0.159 ms
small  | 1 | 2  | 1  | 64   | 16 | True      | flex_attention    | causal          | 0.090 ms
small  | 1 | 2  | 1  | 64   | 16 | True      | flex_attention    | diag_lastrow    | 0.094 ms
small  | 1 | 2  | 1  | 64   | 16 | True      | flex_attention    | diag            | 0.096 ms
small  | 1 | 2  | 1  | 64   | 16 | False     | flash_attention_2 | none            | 0.063 ms
small  | 1 | 2  | 1  | 64   | 16 | False     | flex_attention    | none            | 0.154 ms
small  | 1 | 2  | 1  | 64   | 16 | False     | flex_attention    | causal          | 0.089 ms
small  | 1 | 2  | 1  | 64   | 16 | False     | flex_attention    | diag_lastrow    | 0.088 ms
small  | 1 | 2  | 1  | 64   | 16 | False     | flex_attention    | diag            | 0.087 ms
medium | 1 | 8  | 4  | 1024 | 64 | True      | flash_attention_2 | none            | 0.065 ms
medium | 1 | 8  | 4  | 1024 | 64 | True      | flex_attention    | none            | 0.147 ms
medium | 1 | 8  | 4  | 1024 | 64 | True      | flex_attention    | causal          | 0.177 ms
medium | 1 | 8  | 4  | 1024 | 64 | True      | flex_attention    | diag_lastrow    | 0.206 ms
medium | 1 | 8  | 4  | 1024 | 64 | True      | flex_attention    | diag            | 0.132 ms
medium | 1 | 8  | 4  | 1024 | 64 | False     | flash_attention_2 | none            | 0.066 ms
medium | 1 | 8  | 4  | 1024 | 64 | False     | flex_attention    | none            | 0.145 ms
medium | 1 | 8  | 4  | 1024 | 64 | False     | flex_attention    | causal          | 0.174 ms
medium | 1 | 8  | 4  | 1024 | 64 | False     | flex_attention    | diag_lastrow    | 0.204 ms
medium | 1 | 8  | 4  | 1024 | 64 | False     | flex_attention    | diag            | 0.132 ms
large  | 1 | 16 | 8  | 4096 | 64 | True      | flash_attention_2 | none            | 0.189 ms
large  | 1 | 16 | 8  | 4096 | 64 | True      | flex_attention    | none            | 6.912 ms
large  | 1 | 16 | 8  | 4096 | 64 | True      | flex_attention    | causal          | 6.933 ms
large  | 1 | 16 | 8  | 4096 | 64 | True      | flex_attention    | diag_lastrow    | 6.986 ms
large  | 1 | 16 | 8  | 4096 | 64 | True      | flex_attention    | diag            | 6.865 ms
large  | 1 | 16 | 8  | 4096 | 64 | False     | flash_attention_2 | none            | 0.249 ms
large  | 1 | 16 | 8  | 4096 | 64 | False     | flex_attention    | none            | 7.032 ms
large  | 1 | 16 | 8  | 4096 | 64 | False     | flex_attention    | causal          | 6.807 ms
large  | 1 | 16 | 8  | 4096 | 64 | False     | flex_attention    | diag_lastrow    | 7.055 ms
large  | 1 | 16 | 8  | 4096 | 64 | False     | flex_attention    | diag            | 6.996 ms
```

## Questions
1. Am I doing something wrong?
2. Is this performance difference expected?

Thanks!

**Comment by drisspg** [comment]
Just seeing this but you need to torch.compile(flex_attention) to actually get the correct kernel. there should be an warning thrown for this now
